sdk/dotnet/Pulumi.Automation/XStack.cs

```diff
@@ -0,0 +1,631 @@
+            {
+                var secret = maskedOutput[key] is string maskedValue && maskedValue == "[secret]";
+                output[key] = new OutputValue(value, secret);
+            }
+
+            return output.ToImmutableDictionary();
+        }
+
+        /// <summary>
+        /// Returns a list summarizing all previews and current results from Stack lifecycle operations (up/preview/refresh/destroy).
+        /// </summary>
+        public async Task<ImmutableList<UpdateSummary>> GetHistoryAsync(CancellationToken cancellationToken = default)
+        {
+            var result = await this.RunCommandAsync(new[] { "history", "--json", "--show-secrets" }, null, cancellationToken).ConfigureAwait(false);
+            var options = LocalSerializer.BuildJsonSerializerOptions();
+            var list = JsonSerializer.Deserialize<List<UpdateSummary>>(result.StandardOutput, options);
+            return list.ToImmutableList();
+        }
+
+        public async Task<UpdateSummary?> GetInfoAsync(CancellationToken cancellationToken = default)
+        {
+            var history = await this.GetHistoryAsync(cancellationToken).ConfigureAwait(false);
+            return history.FirstOrDefault();
+        }
+
+        private Task<CommandResult> RunCommandAsync(
+            IEnumerable<string> args,
+            Action<string>? onOutput,
+            CancellationToken cancellationToken)
+            => this.Workspace.RunStackCommandAsync(this.Name, args, onOutput, cancellationToken);
+
+        public void Dispose()
+            => this.Workspace.Dispose();
+
+        private static class ExecKind
+        {
+            public const string Local = "auto.local";
+            public const string Inline = "auto.inline";
+        }
+
+        private enum StackInitMode // TODO: change name of this as per XStack
+        {
+            Create,
+            Select,
+            CreateOrSelect
+        }
+
+        private class InlineLanguageHost : IAsyncDisposable
+        {
+            private readonly TaskCompletionSource<int> _portTcs = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+            private readonly CancellationToken _cancelToken;
+            private readonly IHost _host;
+            private readonly CancellationTokenRegistration _portRegistration;
+
+            public InlineLanguageHost(
+                PulumiFn program,
+                CancellationToken cancellationToken)
+            {
+                this._cancelToken = cancellationToken;
+                this._host = Host.CreateDefaultBuilder()
+                    .ConfigureWebHostDefaults(webBuilder =>
+                    {
+                        webBuilder
+                            .ConfigureKestrel(kestrelOptions =>
+                            {
+                                kestrelOptions.Listen(IPAddress.Any, 0, listenOptions =>
+                                {
+                                    // TODO: HTTPS?
```

Comment:
What's the plan for TODOs? Will we open follow-up issues?

Reply:
I think that's up to you guys. I put a question mark after this one and the logging one because I didn't know if you guys would think those were actually a goal or not. Do we want our language host traffic to go over HTTPS? Do we want to allow consumers to hook into the logging on the language host? If not we can get rid of those TODOs.

Reply:
We can get rid of the todos. Other languages run gRPC services over HTTP on localhost, no https yet.